Src/FluentAssertions/Common/TypeMemberReflector.cs

```diff
@@ -0,0 +1,163 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using FluentAssertions.Equivalency;
+
+namespace FluentAssertions.Common;
+
+/// <summary>
+/// Helper class to get all the public and internal fields and properties from a type.
+/// </summary>
+internal sealed class TypeMemberReflector
+{
+    private const BindingFlags AllInstanceMembersFlag =
+        BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance;
+
+    private readonly Type typeToReflect;
+    private readonly MemberVisibility visibility;
+
+    public TypeMemberReflector(Type typeToReflect, MemberVisibility visibility)
+    {
+        this.typeToReflect = typeToReflect;
+        this.visibility = visibility;
+
+        LoadNonPrivateProperties();
+        LoadNonPrivateFields();
+
+        NonPrivateMembers = NonPrivateProperties.Concat<MemberInfo>(NonPrivateFields).ToArray();
+    }
+
+    public MemberInfo[] NonPrivateMembers { get; }
+
+    public PropertyInfo[] NonPrivateProperties { get; private set; }
+
+    public FieldInfo[] NonPrivateFields { get; private set; }
+
+    private void LoadNonPrivateProperties()
+    {
+        IEnumerable<PropertyInfo> query =
+            from propertyInfo in GetPropertiesFromHierarchy(typeToReflect, visibility)
+            where HasNonPrivateGetter(propertyInfo)
+            where !propertyInfo.IsIndexer()
+            select propertyInfo;
+
+        NonPrivateProperties = query.ToArray();
+    }
```

Comment:
If we return the `PropertyInfo[]` here and assign it to `NonPrivateProperties` in the constructor, then:
* `NonPrivateProperties` can be `{ get }` ,
* `LoadNonPrivateProperties` can be `static` with `typeToReflect` and `visibility` as parameters  and
* `typeToReflect` and `visibility` are no longer need as members on this class.


In general I strive to only call static methods from constructors to avoid risking using members before they have been initialized.

Reply:
I think it's a matter of taste, and I kind of liked having those two fields. But I have adopted your suggestions anyhow.